packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -895,6 +1080,687 @@ void Add(
+  KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const
+  {
+    //count the union of nonzeros in Arow and Brow
+    Ordinal numEntries = 0;
+    Ordinal ai = 0;
+    Ordinal bi = 0;
+    auto Arowstart = Arowptrs(i);
+    auto Arowlen = Arowptrs(i + 1) - Arowstart;
+    auto Browstart = Browptrs(i);
+    auto Browlen = Browptrs(i + 1) - Browstart;
+    while(ai < Arowlen && bi < Browlen)
+    {
+      //have an entry in C's row
+      numEntries++;
+      auto Acol = Acolinds(Arowstart + ai);
+      auto Bcol = Bcolinds(Browstart + bi);
+      if(Acol <= Bcol)
+        ai++;
+      if(Acol >= Bcol)
+        bi++;
+    }
+    numEntries += Arowlen - ai;
+    numEntries += Browlen - bi;
+    Crowcounts(i) = numEntries;
+  }
+  const RowPtrs& Arowptrs;
+  const ColInds& Acolinds;
+  const RowPtrs& Browptrs;
+  const ColInds& Bcolinds;
+  RowPtrs& Crowcounts;
+};
+
+template<typename Scalar, typename Ordinal, typename Values, typename RowPtrs, typename ColInds>
+struct SumFunctor
+{
+  SumFunctor(const Values& Avals_, const RowPtrs& Arowptrs_, const ColInds& Acolinds_, const Scalar& scalarA_,
+           const Values& Bvals_, const RowPtrs& Browptrs_, const ColInds& Bcolinds_, const Scalar& scalarB_,
+           Values& Cvals_, const RowPtrs& Crowptrs_, ColInds& Ccolinds_) :
+    Avals(Avals_), Arowptrs(Arowptrs_), Acolinds(Acolinds_), scalarA(scalarA_),
+    Bvals(Bvals_), Browptrs(Browptrs_), Bcolinds(Bcolinds_), scalarB(scalarB_),
+    Cvals(Cvals_), Crowptrs(Crowptrs_), Ccolinds(Ccolinds_) {}
+  KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const
+  {
+    //get views of row i in A and B
+    auto Arowstart = Arowptrs(i);
+    auto Arowlen = Arowptrs(i + 1) - Arowstart;
+    auto Browstart = Browptrs(i);
+    auto Browlen = Browptrs(i + 1) - Browstart;
+    Ordinal CrowStart = Crowptrs(i);
+    Ordinal ai = 0;
+    Ordinal bi = 0;
+    Ordinal numInserted = 0;
+    while(ai < Arowlen && bi < Browlen)
+    {
+      Ordinal Acol = Acolinds(Arowstart + ai);
+      Ordinal Bcol = Bcolinds(Browstart + bi);
+      //next col in C is minimum of two next cols in A and B
+      Ordinal Ccol = Acol < Bcol ? Acol : Bcol;
+      Scalar val = 0;
+      if(Acol == Ccol)
+      {
+        val += scalarA * Avals(Arowstart + ai);
+        ai++;
+      }
+      if(Bcol == Ccol)
+      {
+        val += scalarB * Bvals(Browstart + bi);
+        bi++;
+      }
+      Cvals(CrowStart + numInserted) = val;
+      Ccolinds(CrowStart + numInserted) = Ccol;
+      numInserted++;
+    }
+    for(; ai < Arowlen; ai++)
+    {
+      Cvals(CrowStart + numInserted) = scalarA * Avals(Arowstart + ai);
+      Ccolinds(CrowStart + numInserted) = Acolinds(Arowstart + ai);
+      numInserted++;
+    }
+    for(; bi < Browlen; bi++)
+    {
+      Cvals(CrowStart + numInserted) = scalarB * Bvals(Browstart + bi);
+      Ccolinds(CrowStart + numInserted) = Bcolinds(Browstart + bi);
+      numInserted++;
+    }
+  }
+  const Values& Avals;
```

Comment:
See above note on Kokkos::View needing to be stored by value in functors, not by reference.